tmt/hardware.py

```diff
@@ -0,0 +1,1088 @@
+        ) -> Callable[[SpecType, int], ConstraintBase]:
+    @functools.wraps(fn)
+    def wrapper(spec: SpecType, index: int) -> ConstraintBase:
+        constraint = fn(spec, index)
+
+        if isinstance(constraint, CompoundConstraint) and len(constraint.constraints) == 1:
+            return constraint.constraints[0]
+
+        return constraint
+
+    return wrapper
+
+
+@ungroupify
+def _parse_boot(spec: SpecType) -> ConstraintBase:
+    """
+    Parse a boot-related constraints.
+
+    :param spec: raw constraint block specification.
+    :returns: block representation as :py:class:`ConstraintBase` or one of its subclasses.
+    """
+
+    group = And()
+
+    if 'method' in spec:
+        constraint = Constraint.from_specification(
+            'boot.method', spec["method"], as_quantity=False)
+
+        if constraint.operator == Operator.EQ:
+            constraint.change_operator(Operator.CONTAINS)
+
+        elif constraint.operator == Operator.NEQ:
+            constraint.change_operator(Operator.NOTCONTAINS)
+
+        group.constraints += [constraint]
+
+    return group
+
+
+@ungroupify
+def _parse_virtualization(spec: SpecType) -> ConstraintBase:
+    """
+    Parse a virtualization-related constraints.
+
+    :param spec: raw constraint block specification.
+    :returns: block representation as :py:class:`ConstraintBase` or one of its subclasses.
+    """
+
+    group = And()
+
+    if 'is-virtualized' in spec:
+        group.constraints += [
+            Constraint.from_specification(
+                'virtualization.is_virtualized',
+                str(spec['is-virtualized']),
+                as_quantity=False,
+                as_cast=bool
+                )
+            ]
+
+    if 'is-supported' in spec:
+        group.constraints += [
+            Constraint.from_specification(
+                'virtualization.is_supported',
+                str(spec['is-supported']),
+                as_quantity=False,
+                as_cast=bool
+                )
+            ]
+
+    if 'hypervisor' in spec:
+        group.constraints += [
+            Constraint.from_specification(
+                'virtualization.hypervisor',
+                spec['hypervisor'],
+                as_quantity=False
+                )
+            ]
+
+    return group
+
+
+@ungroupify
+def _parse_compatible(spec: SpecType) -> ConstraintBase:
+    """
+    Parse constraints related to the compatible distro parameter.
+
+    :param spec: raw constraint block specification.
+    :returns: block representation as :py:class:`ConstraintBase` or one of its subclasses.
+    """
+
+    group = And()
+
+    for distro in spec.get('distro', []):
+        constraint = Constraint.from_specification('compatible.distro', distro, as_quantity=False)
+
+        constraint.change_operator(Operator.CONTAINS)
+
+        group.constraints.append(constraint)
+
+    if len(group.constraints) == 1:
+        return group.constraints[0]
```

Comment:
Isn't this covered by `ungroupify` already?

Reply:
Fixed in e7c0774a.